### ISSUE TITLE
Add OpenStack Support

### ISSUE THREAD

**Issue by dgoodwin** [openshift/hive#897]
Opening a new issue specifically for OpenStack to help with public coordination. (similar to #889)

@joelddiaz is working on this and will have a PR up very shortly, main issue right now is testing. 

Joel can you summarize the caveat that will be present in the first draft where we will not be automating DNS and floating IPs / app ingress?

CC @rasheedamir @waseem-h @gyliu513

**Comment by joelddiaz** [comment]
Yes, I'll also put this info in the PR as well.

The OpenShift IPI installer (and UPI for that matter) assumes a certain amount of DNS setup has occurred before the installation is launched. This means, you will create a Floating IP and make the appropriate DNS entry to point to the Floating IP (the FIP will be passed into the installer).

During installation, the ingress/apps setup also does not cover DNS, so post-installation you need to create an FIP, make the DNS entry for it, and set/bind the FIP to the OpenStack port that was created during installation.

TL;DR:
1) openstack floating ip create <network>
2) make DNS A record to point to the FIP (eg api.mycluster.mydomain.com -> FIP)
3) run installer/hive
4) create another FIP, bind it to the ingress port
```
INGRESS_PORT=$(openstack port list --format value -c Name | awk "/${CLUSTER_NAME}.*-ingress-port/ {print}")
openstack floating ip set --port ${INGRESS_PORT} ${APPS_FIP}
```
5) DNS entry for the ingress/apps (eg *.apps.mycluster.mydomain.com -> INGRESS_FIP)

**Comment by joelddiaz** [comment]
and FWIW, the DNS setup and floating IP creation can all happen before install. it is only the final 'bind FIP to ingress port' that must be done post-install

**Comment by rasheedamir** [comment]
@joelddiaz 

- will it work seamlessly for UPI? 
- when will you make the PR?

**Comment by joelddiaz** [comment]
We don't presently target UPI workflows (the baremetal support might be considered some kind of UPI... I'm not very familiar with the baremetal work), although this OpenStack effort is the most manual installation effort (of the currently supported clouds) simply because of the DNS steps. We'll have to think about how we want DNS to work long term for this workflow.
I expect to have the PR up tomorrow.

**Comment by rasheedamir** [comment]
ok; we haven't been looking into IPI due to it's hard dependency on Swift; but the great news is that 4.4 is going to be Swift'less; so, which means we can try out IPI as well; [4.4.0-rc.4](https://mirror.openshift.com/pub/openshift-v4/clients/ocp/4.4.0-rc.4/) is available and expectation is that 4.4 will be released by end of April; so, it's pretty near.

also our target is to have full automation; so, that other dependencies like DNS can be setup via terraform if required before hand!

we are looking forward to PR! and collaborating on testing it out

**Comment by dgoodwin** [comment]
Regarding automating the DNS, in our discussions thus far I believe we assumed that for the overall API DNS and FIP, something configures this before creating the ClusterDeployment, and then provides the info *in* the ClusterDeployment spec, should be visible shortly in Joel's PR.

For the ingress DNS and FIP, this could easily be automated by a separate operator watching ClusterDeployment with the specific users implementation for setting this up.

**Comment by rasheedamir** [comment]
> something configures this before creating the ClusterDeployment,

so, what is the recommendation or suggestion for this "something" 😂? if one wants seamless experience.

> For the ingress DNS and FIP, this could easily be automated by a separate operator watching ClusterDeployment with the specific users implementation for setting this up.

Agreed and makes sense

**Comment by dgoodwin** [comment]
> > something configures this before creating the ClusterDeployment,
> 
> so, what is the recommendation or suggestion for this "something" joy? if one wants seamless experience.

In our main deployments (OpenShift Dedicated) there is a frontend orchestrating the provisioning of clusters, it does things first like obtain a cloud provider account for the cluster (using a CRD and a separate account operator to fulfill the request), *then* it creates the Hive ClusterDeployment with the account details allowing us to start provisioning. We would expect the same here.

However we could also talk about a more formal pre-req mechanism if desired.

**Comment by dgoodwin** [comment]
Pr #902

**Comment by tobru** [comment]
> ... using a CRD and a separate account operator to fulfill the request

That's interesting. Is this account operator Open Source as well? If yes, where can I find it?

**Comment by tobru** [comment]
Answering myself: Seems to be this one https://github.com/openshift/aws-account-operator

**Comment by gregsheremeta** [comment]
this is done
